Show a pointer/hint in the settings tab informing the user about the styles tab

talldan opened the pull request:
## What?
Fixes #47410

## Why?
It's nice to let the user know about this change

## How?
- Implements a new (non-public, non-reusable) component for showing this pointer ([see the discussion in the issue about why `Notice` wasn't used](https://github.com/WordPress/gutenberg/issues/47410#issuecomment-1407969824))
- Uses the preferences package to record the state of whether the notice has been dismissed or not.

## Testing Instructions
Note - if you want to test this multiple times, you can run this snippet in the browser console to bring the dismissed pointer back again:
```js
wp.data.dispatch( 'core/preferences' ).set( 'core', 'isInspectorControlsTabsHintVisible', true )
```

1. Add a block (a good one to test with is 'Group')
2. Observe the 'pointer' in the sidebar
3. Dismiss it (at this point it shouldn't ever come back again)

### Testing Instructions for Keyboard
1. Add a group block by tabbing into the editor content and typing `/group` and hitting the enter key
2. Tab to the 'Editor settings' region. If it's not open you should land upon a button 'Open document settings' first, so use that button to open it.
3. Tab till you get to a 'Dismiss hint' button, optionally read the text that's in the DOM before that button (but not focusable).
4. Select that button.
5. Focus should be transferred to the Advanced section as the hint disappears

## Screenshots or screencast <!-- if applicable -->
![Screen Shot 2023-02-02 at 1 21 21 pm](https://user-images.githubusercontent.com/677833/216238609-05ecf77f-1c1a-44c3-b840-1fa5280ff30f.png)

talldan committed:
Show a pointer/hint in the settings tab informing the user about the styles tab

talldan committed:
Add a spoken message when the notice is dismissed

talldan commented:
I could do with some design advice:
- Show the text "They've moved to the styles tab" be on a new line?
- Do you want any kind of animation when dismissing the pointer?

And also accessibility feedback would be good. I'll copy what I posted in the description:
- I could do with some accessibility feedback on this, as at the moment there's a focusable 'Dismiss' button without much context around what it dismisses. What's the best way to add context?
- Selecting the 'dismiss' button causes a focus loss, as the button removes itself from the DOM. Is it worth transferring focus to an adjacent element?

aaronrobertshaw reviewed (commented):
Thanks for picking this one up @talldan 👍 

Potential design and a11y tweaks aside, this one is testing as advertised for me. Both via the mouse or keyboard navigation. 

Appreciate the snippet to reset the preference as well, that saved me some time 🙇 

> Show the text "They've moved to the styles tab" be on a new line?

This would be a little neater in my opinion but for a language other than English it might lead to a worse outcome. 

> Do you want any kind of animation when dismissing the pointer?

This could be a follow-up task if there is some time pressure to land this early to help introduce users to the tabs.


https://user-images.githubusercontent.com/60436221/216248144-016e4ba4-78dd-46f3-8632-453f5c7abcfa.mp4

jasmussen reviewed (approved):
Nice, thanks for doing that!

A few very nitpicky details, then this looks great:

* Set `border-radius: $radius-block-ui;` on the gray container
* Set the padding to `$grid-unit-15` instead of `$grid-unit-20`
* You might need to fiddle the dismiss button paddings down to the 05 unit to match

joedolson commented:
Context: Yeah, that's a bit difficult. In reading mode, a screen reader user would read the text and then find the dismiss button, but if they were navigating by tabs, they wouldn't. I wouldn't go overboard with providing context: something as simple as "Dismiss Notice" is enough to let a user know that they should explore the notice and let them know that this isn't a control to dismiss a control panel, etc. I favor lower verbosity, so that's what I would choose to do. @alexstine might have different preferences.

Focus movement: focus should move to the nearest focusable point. Usually, that would be the next focusable element; but because this is at the end of the settings panel, I think it would make more sense to move the user to the previous focusable element, which should be the 'Advanced' settings toggle.  The next focusable is the link to move to the currently selected block, and that would also be fine, though more surprising to sighted mouse users, who wouldn't normally see that control.

Have you considered making the text 'styles tab' a link that can navigate the user into the styles tab? If a user does encounter this notice, it could be because they are looking for those settings, and offering an easy path to reach them immediately would be useful.

aaronrobertshaw commented:
> Have you considered making the text 'styles tab' a link that can navigate the user into the styles tab?

@joedolson this would be a nice touch. 

Unfortunately, at the moment the `TabPanel` component doesn't allow for programmatic switching of the selected tab. A refactor for that component has been discussed previously so hopefully we can achieve something like this in the not too distant future.

alexstine commented on packages/block-editor/src/components/inspector-controls-tabs/settings-tab-pointer.js:
```diff
@@ -0,0 +1,45 @@
+				label={ __( 'Dismiss' ) }
+				onClick={ () => {
+					setPreference( 'core', PREFERENCE_NAME, false );
+					speak( __( 'Notice dismissed.' ) );
```
I think this speak is not needed as long as you manage the focus. I think the context is good enough here with the label itself for a user to know what happened. The only time I really worry about notifying users is if it is urgent information/important information. Examples might include submitting a form and receiving a success/error message or providing navigation instructions in a non-native element, this one is closely related to Gutenberg select/edit modes.

alexstine reviewed (commented):
@talldan I agree with Joe, I think previous element to focus is okay and this makes sense due to the tab order. Someone would tab on advanced first before receiving the dismiss button.

Since this notice is not overly important, I do not believe we should add verbosity. There is an argument to be made though on how useful this notice would be at all considering a user would have to tab through the entire inspector to reach it. Before the new tabs, didn't the styles section appear before the advanced options? Might be worth moving that notice directly to the position where those style settings would have appeared in the DOM.

EDIT: One other note, can you think of something else to call this besides a settings pointer? Maybe a settings notice pointer? Looking at the code, the import statement, I would have no idea what this component is.

talldan committed:
Update margin around text and button

talldan committed:
Remove spoken message

talldan committed:
Rename to hint and add focus management code

talldan committed:
Relabel dismiss button

talldan commented:
Thanks for all the feedback. It was incredibly useful. I've made some changes:
- Updates the styles as recommended
- Renamed the component from `Pointer` to `Hint` (I agree, pointer isn't the best given the existing programming and UI meanings for that word).
- Removed the spoken message on pressing the dismiss button
- Move focus to the previous element when the dismiss button is pressed
- Updated the button label from 'Dismiss' to 'Dismiss hint', so there's some clarity over what it dismisses.

> Might be worth moving that notice directly to the position where those style settings would have appeared in the DOM.

@alexstine That's tricky to do as there's no space, and on the other hand there's lots of empty space after Advanced. Given the low prominence of this message, I think this is the right trade-off.

alexstine commented:
@talldan Fair enough. I just wanted to make sure it was not too undiscoverable for people who could not see it all the way down there. No reason to worry though I guess if it is not meant to be an essential message.

I will try to give this a test this weekend or maybe tomorrow. Really behind on things after a winter storm ripped through Texas.

Thanks.

talldan commented:
Thanks Alex, hope the storm hasn't caused you too many issues.

aaronrobertshaw reviewed (approved):
Thanks for addressing all the feedback @talldan 🚀 

The recent changes look good and test as advertised. In particular, the focus management and moving focus to the previous element when the hint is dismissed works well.

I'm happy to approve this now but waiting for Alex's testing and feedback might be wise. To make that a little easier, perhaps we could update the PR's description and test instructions to reflect the current state of the PR?

https://user-images.githubusercontent.com/60436221/216537037-7aa49ea5-810c-43f1-bd71-3dd87ba7bb1f.mp4

talldan commented:
> To make that a little easier, perhaps we could update the PR's description and test instructions to reflect the current state of the PR?

Oh yep, thanks for the reminder 👍

alexstine reviewed (approved):
@talldan Perfection. Works well and code looks good.

richtabor commented:
Good job team!

ntsekouras commented:
Cherry-picked this PR to the wp/6.2 branch.
